common/src/main/java/com/wynntils/functions/SpellFunctions.java

```diff
@@ -0,0 +1,139 @@
+                    new FunctionArguments.Argument<>("isColored", Boolean.class, true),
+                    new FunctionArguments.Argument<>("useShortName", Boolean.class, false)));
+        }
+    }
+
+    public static class ShamanTotemStateFunction extends Function<String> {
+
+        @Override
+        public String getValue(FunctionArguments arguments) {
+            int totemNumber = arguments.getArgument("totemNumber").getIntegerValue();
+
+            ShamanTotem shamanTotem = Models.ShamanTotem.getTotem(totemNumber);
+
+            if (shamanTotem == null) {
+                return "";
+            }
+
+            return shamanTotem.getState().toString().toUpperCase(Locale.ROOT);
+        }
+
+        @Override
+        public FunctionArguments.Builder getArgumentsBuilder() {
+            return new FunctionArguments.RequiredArgumentBuilder(
+                    List.of(new FunctionArguments.Argument<>("totemNumber", Integer.class, null)));
+        }
+    }
+
+    public static class ShamanTotemLocationFunction extends Function<String> {
+
+        @Override
+        public String getValue(FunctionArguments arguments) {
+            int totemNumber = arguments.getArgument("totemNumber").getIntegerValue();
+
+            ShamanTotem shamanTotem = Models.ShamanTotem.getTotem(totemNumber);
+
+            if (shamanTotem == null) {
+                return "";
+            }
+
+            return shamanTotem.getLocation().toString();
+        }
+
+        @Override
+        public FunctionArguments.Builder getArgumentsBuilder() {
+            return new FunctionArguments.RequiredArgumentBuilder(
+                    List.of(new FunctionArguments.Argument<>("totemNumber", Integer.class, null)));
+        }
+    }
+
+    public static class ShamanTotemTimeLeftFunction extends Function<Integer> {
+
+        @Override
+        public Integer getValue(FunctionArguments arguments) {
+            int totemNumber = arguments.getArgument("totemNumber").getIntegerValue();
+
+            ShamanTotem shamanTotem = Models.ShamanTotem.getTotem(totemNumber);
+
+            if (shamanTotem == null) {
+                return 0;
+            }
+
+            return shamanTotem.getTime();
+        }
+
+        @Override
+        public FunctionArguments.Builder getArgumentsBuilder() {
+            return new FunctionArguments.RequiredArgumentBuilder(
+                    List.of(new FunctionArguments.Argument<>("totemNumber", Integer.class, null)));
+        }
+    }
+
+    public static class ShamanTotemDistanceFunction extends Function<Double> {
+        @Override
+        public Double getValue(FunctionArguments arguments) {
+            int totemNumber = arguments.getArgument("totemNumber").getIntegerValue();
+
+            ShamanTotem shamanTotem = Models.ShamanTotem.getTotem(totemNumber);
+
+            if (shamanTotem == null) {
+                return 0d;
+            }
+
+            return McUtils.player()
```

Comment:
As a future improvement, it would be cool to have a Location or Coord type, and then you could apply a `distance` function to it. Then we could use that to show distance to compass beam etc as well. And we could introduce stuff like `travel_time`, taking into account distance and current avg bps, etc :-)

Reply:
Agree.